### PR TITLE
Add new AtU8 beam chunk

### PR DESCRIPTION
**Pull request by josevalim**
The new chunk stores atoms encoded in UTF-8.

This is still work in progress and the following tasks are still pending:
- [x] Add the compile option `r19` that will compile atoms to the old "Atom" chunk [as mentioned by @bjorng in the mailing list](http://erlang.org/pipermail/erlang-questions/2016-February/087509.html)
- [x] [Support 255 UTF-8 codepoints in `binary_to_atom` instead of 255 bytes](https://github.com/josevalim/otp/blob/jv-atu8-chunk/erts/emulator/beam/erl_unicode.c#L1900). The issue is that calculating the number of characters is linear to the binary size. Although `erts_atom_put` performs such calculation, it returns a NON-VALUE for both invalid encoding and large binaries errors, and [the test suite expects badarg for invalid encoding and system limit for large binaries](https://github.com/josevalim/otp/blob/jv-atu8-chunk/erts/emulator/test/bif_SUITE.erl#L458). To solve this, we can either:
  1. add a new function, used by both erts_atom_put and binary_to_atom, that properly encodes the two kinds of errors so we can act accordingly
  2. change binary_to_atom to only raise badarg and no longer raise on system_limit
  3. compute the size in the `binary_to_atom` BIF and continue raising a system limit error (effectively traversing the binary twice)
  
  Thoughts? Any other options? Once this is changed, the docs shall be properly updated.

Feedback request:
1. `beam_lib` has been modified to support a new 'utf8_atoms' chunk that maps to the new "AtU8" chunk. This means accessing the old "Atom" chunk can now be missing although it is straight-forward to handle it with a case statement.
2. `list_to_atom` has not been modified and it will still fail if given a character more than 255. The reason I have decided to not change `list_to_atom` is because its current documentation does not say it may support characters more than 255 in the future (while the `binary_to_atom` has a very explicit warning about such).

### CODE REVIEW

**Comment by ferd** [comment]
What is meant by an UTF8 codepoint? I thought that by the time you're in UTF-8 you no longer deal in codepoints, just in bytes?

The Unicode consortium does specify that lengths can be counted in one of 4 ways. From the [FAQ](http://unicode.org/faq/char_combmark.html) using the string aनि亜𐂃 (U+0061, U+0928, U+093F, U+4E9C, U+10083)
- bytes: UTF-8: 14, UTF-16:  12, UTF-32: 20
- code units (position in a string with the given size): UTF-8: 14, UTF-16: 6, UTF-32: 5
- code points (how many unicode code points are in there): 5 for all, since there is a combining mark
- grapheme clusters: 4 for all of them, since this is a logical grouping for the length based on 'characters' as we people usually think of them, rather than the implementation of them.

The two latter are not encoding-specific. What is meant exactly by UTF-8 code points? Just code points, or it's a bad term for code units (and hence bytes)?

**Comment by josevalim** [comment]
Thanks @ferd. I meant literally code points as in your definition:

> code points (how many unicode code points are in there): 5 for all, since there is a combining mark

It is also worth mentioning the implementation today (prior to this patch) also checks code points. This can be [checked in the source code](https://github.com/erlang/otp/blob/e1489c448b7486cdcfec6a89fea238d88e6ce2f3/erts/emulator/beam/erl_unicode.c#L1158) or by trying a quick example:

```erl
3> binary_to_atom(binary:copy(<<"é"/utf8>>, 255), utf8).
ééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééé
4> byte_size(binary:copy(<<"é"/utf8>>, 255)).
510
5> binary_to_atom(binary:copy(<<"é"/utf8>>, 256), utf8).
** exception error: a system limit has been reached
     in function  binary_to_atom/2
        called as binary_to_atom(<<"éééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééééé"/utf8...>>,
                                 utf8)
```

So this pull request is currently backwards incompatible. The question is mostly how to implement the new binary_to_atom efficiently.

File: erts/emulator/beam/beam_load.c [hunk 1]
```diff
@@ -6332,7 +6355,7 @@ erts_make_stub_module(Process* p, Eterm Mod, Eterm Beam, Eterm Info)
 	goto error;
     }
     define_file(stp, "atom table", ATOM_CHUNK);
-    if (!load_atom_table(stp)) {
+    if (!load_atom_table(stp, ERTS_ATOM_ENC_LATIN1)) {
```
Comment (margnus1) [inline erts/emulator/beam/beam_load.c]: You'll need to go through the same `if (stp->chunks[UTF8_ATOM_CHUNK].size > 0) ...` charade down here, or `code:make_stub_module/3` (and thus HiPE) will break. There's some tests for it in `erts/emulator/test/code_SUITE*`

Comment (josevalim) [inline erts/emulator/beam/beam_load.c]: Thank you. I have run the code suite and fixed this and the remaining failing tests.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by bjorng** [comment]
We will not have time to do a thorough review until OTP 19.0 has been released, but here is some quick feedback:

The test case `lc_SUITE:effect/1` fails.

`binary_to_atom()`: Solution i, that is keep the existing error semantics and implement it efficiently.

Not modifying `list_to_atom/1` is an interesting idea that we have not considered.

**Comment by nox** [comment]
@bjorng Can we name the chunk 'JOSÉ'? PLS.

**Comment by bjorng** [comment]
We will take a closer at this PR request when you have fixed the failed test case mentioned above and rebased it on the latest master.

**Comment by nox** [comment]
> It is also worth mentioning the implementation today (prior to this patch) also checks code points. This can be checked in the source code or by trying a quick example:

I think you interpreted that incorrectly. Currently only UTF-8 encoded atoms that can be converted to Latin-1 are supported, so it just counts the number of Latin-1 code points, which equates the number of Latin-1 code units, which equates the number of bytes in the result.

**Comment by josevalim** [comment]
@bjorng I will allocate time in the next weeks to work on all of my pending pull requests. :D

While talking to @nox, we came up with another solution for the second bullet above. @nox commented that it would be much simpler to count bytes instead of counting codepoints. However, if we count bytes and keep the limit of 255 bytes, it would be backwards incompatible because an atom made only of the letter `é` in latin1 will now take double the space in UTF-8, possibly triggering the system limit.

However, if we double the maximum atom size, we could keep the conceptually simpler and faster code that only counts bytes and remain backwards compatible.

From the Elixir perspective, Increasing the atom limit would be a positive change, since function names  must be atoms, sometimes we may reach this limit when building functions dynamically.

@bjorng thoughts on this suggestion?

**Comment by bjorng** [comment]
I think that the BIFs that create atoms should count code points, and let the limit be 255 code points. When converting from a list, the number of code points is the number of elements in the list (assuming that we don't try to combine code points that can be combined).

The chunk in the BEAM file can give the size of each atom in bytes, and the loader can trust the compiler not to create overlong atoms. (There could be a sanity check to abort the loading if any atom is longer than 1024 bytes.)

**Comment by bjorng** [comment]
Rationale for counting code points: We don't want to handle support request from Klingon users that complain that they can only put 85 Klingon characters into their atoms, while speakers of most west-European languages can have 255 characters in their atoms.

**Comment by nox** [comment]
What about following what ROK proposed in EEP20?

2 bytes for byte length, 2 bytes for code point count.

File: erts/emulator/beam/atom.c [hunk 2]
```diff
@@ -233,10 +233,10 @@ static void latin1_to_utf8(byte* conv_buf, const byte** srcp, int* lenp)
 }
 
 /*
- * erts_atom_put() may fail. If it fails THE_NON_VALUE is returned!
+ * erts_atom_put_index() may fail. Returns -1 for bad encoding, -2 for system limit.
  */
-Eterm
-erts_atom_put(const byte *name, int len, ErtsAtomEncoding enc, int trunc)
+int
+erts_atom_put_index(const byte *name, int len, ErtsAtomEncoding enc, int trunc)
```
Comment (josevalim) [inline erts/emulator/beam/atom.c]: @bjorng I have introduced this function that returns the atom index instead of the atom so we have control upstream if we should error with badarg or system_limit. I have two questions I would love your input on:

1. Should the name have the `erts_` prefix?

2. Should we use `#define ATOM_BAD_ENCODING` and `#define ATOM_MAX_CHARS` instead of using integers? If so, is it fine to define those on atom.h?

Notice the new function and its return values are used in binary_to_atom (erl_unicode.c). The PR has also been rebased. Once those questions are answered, I will squash everything and we should be good to go.

Comment (psyeugenic) [inline erts/emulator/beam/atom.c]: The convention for the `erts_` prefix is to use it for global functions. Older function, or some very often used functions, e.g. `size_object` does not use it though.

Comment (bjorng) [inline erts/emulator/beam/atom.c]: 2. Yes, and yes.

**Comment by josevalim** [comment]
@bjorng This is ready for review. I have rebased, fixed all pending feedback and features.

I have ran emulator, compiler and stdlib test suites:

  * emulator: One failure regarding registered sends across nodes (I inspected the code and it seems to not related be to this patch)
  * compiler: No failures
  * stdlib: One failure in the shell suite which seems unrelated

The only pending task is to add a test for the r19 option. However, since we cannot write a utf8 atom using single quotes, I believe we will need to build the Erlang Abstract Format by hand. Unless @bjorng has a better idea, I will include this test in the next 24h.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by bjorng** [comment]
Three test cases fail in `compile_SUITE`: `core`, `core_roundtrip`, and `asm`.

I updated the primary bootstrap and compiled from a clean repository.

**Comment by josevalim** [comment]
> I updated the primary bootstrap and compiled from a clean repository.

How do I update the primary bootstrap? I remember doing a clean build but I did not touch the bootstrap files. Thanks!

**Comment by bjorng** [comment]
Assuming that you have already built the system, this command updates the primary bootstrap:

    PATH=$ERL_TOP/bin:$PATH ./otp_build update_primary

It builds and commit a new primary bootstrap. The script may leave a few files that are modified by not committed. If so, do:

    git commit --amend -a

**Comment by josevalim** [comment]
@bjorng Thanks, I will give it a try.

To be clear, should I include the bootstrap updates as part of my pull request or better not to avoid conflicts?

**Comment by bjorng** [comment]
Don't include the bootstrap.

**Comment by josevalim** [comment]
@bjorng see reports on the failures below.

## asm

The ASM failure happens because we generate the .S file with UTF-8 atoms and then we try to read them back. However, `erl_scan:tokens(...)` does not accept UTF-8 atoms because of this line: https://github.com/erlang/otp/blob/3e06b82f0f29d90bff0783e7f3d1dabb435782f5/lib/stdlib/src/erl_scan.erl#L773

For now, I have decided explicitly to not support char > 255 on `list_to_atom/1`. How should we proceed? Should we allow char > 255 on `list_to_atom/1`? Or should I convert it to a binary and then call `binary_to_atom/2` exclusively on erl_scan?

## core

The other failures are a bit weird. They can be fixed by adding `unicode:characters_to_binary(CorePP)` on the two lines below:

* https://github.com/erlang/otp/blob/3e06b82f0f29d90bff0783e7f3d1dabb435782f5/lib/compiler/test/compile_SUITE.erl#L722

* https://github.com/erlang/otp/blob/3e06b82f0f29d90bff0783e7f3d1dabb435782f5/lib/compiler/test/compile_SUITE.erl#L794

I have also tried fixing it by adding `file:write_file(CoreFile, CorePP, [{encoding, utf8}])` but that was not enough to fix the error (it still returns `{error, badarg}` which I assume due to conversion errors). I have checked CorePP and it does contain valid UTF-8. On which situation `file:write_file/2 + unicode:characters_to_binary/1` would emit a different result than doing the `file:write_file/3` with the encoding option?

**Comment by bjorng** [comment]
> For now, I have decided explicitly to not support char > 255 on list_to_atom/1. How should we proceed? Should we allow char > 255 on list_to_atom/1? Or should I convert it to a binary and then call binary_to_atom/2 exclusively on erl_scan?

I saw that as temporary limitation. Ultimately, we want `list_to_atom/1` to be able to handle all Unicode characters. Therefore, I think you should remove the limitation.

(We will have to review the use of `list_to_atom/1` in the OTP source code before merging this branch.)

**Comment by josevalim** [comment]
> (We will have to review the use of list_to_atom/1 in the OTP source code before merging this branch.)

*runs*

**Comment by bjorng** [comment]
> The other failures are a bit weird. They can be fixed by adding unicode:characters_to_binary(CorePP) on the two lines below:

I think that is good way to fix them.

> I have also tried fixing it by adding file:write_file(CoreFile, CorePP, [{encoding, utf8}]) but that was not enough to fix the error (it still returns {error, badarg} which I assume due to conversion errors). I have checked CorePP and it does contain valid UTF-8. On which situation file:write_file/2 + unicode:characters_to_binary/1 would emit a different result than doing the file:write_file/3 with the encoding option?

I think this is a bug of some sort. I am not sure, but I think the `file:write_file/3` did not exist when handling of different encodings were added. The `encoding` option was probably only intended for IO going through the `io` module, not for the file operations in `file`.

**Comment by bjorng** [comment]
> runs

Come back! :-) I meant members of the OTP team, but others are welcome to help us with it.

**Comment by psyeugenic** [comment]
> > runs

> Come back! :-) I meant members of the OTP team, but others are welcome to help us with it.

*runs*

**Comment by nox** [comment]
> > > runs

> > Come back! :-) I meant members of the OTP team, but others are welcome to help us with it.

> runs

*had ran away already*

**Comment by josevalim** [comment]
@bjorng I started working on the list_to_atom and I am not sure if there already exists a C function that receives a list of integers and converts it to a buffer (byte array) while converting to utf-8 encoding. Well, I am sure there is one that converts a char data to a binary (because of unicode.erl) but I am not sure of an intermediate (list to buffer). Thoughts? Suggestions?

**Comment by bjorng** [comment]
If you will allow me to push additional commits to your pull request, I will push a commit that teaches `list_to_atom/1` to handle Unicode characters. (Tick the check box for "Allow edits from maintainers" in the right-hand column.)

**Comment by josevalim** [comment]
@bjorng done :D

**Comment by bjorng** [comment]
I have pushed a new commit. (Feel free to squash it into your own commit.) I have not tested it thoroughly, only that the run-time system can be started and that a few Russian words are properly converted from strings to atoms.

Please verify that the test cases now pass. Also do a rebase on the latest master.

**Comment by josevalim** [comment]
@bjorng thank you so much!

I have also added Unicode characters support to `list_to_existing_atom/1` and written tests for both `list_to_atom/1` and `list_to_existing_atom/1`. There are now three commits but I can squash them altogether (although I don't want to take credit for your changes). Tests for emulator and compiler are passing locally.

I have also rebased and pushed. :tada:

Finally, during the weekend I took a look at most list_to_atom/atom_to_list calls in OTP and the ones that have called my attention are the ones in the compiler. I will take a double look at the compiler during this week and let you know if I find out anything suspicious.

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Teach list_to_existing_atom/1 to handle all Unicode characters


----------------

I am a script, I am not human 

----------------

**Comment by bjorng** [comment]
Thanks!

I have pushed two commits. One is correction for the `compile_SUITE:kernel_listing/1` test case which is now included because of the rebase. Since I added the test case, it's only fair that I correct it. The other commit adds a unicode string and unicode string to `simple.erl`, which is used by the `listings/1` test case.

**Comment by bjorng** [comment]
Regarding credit, it is sufficient that you credit me in the commit message. I care much more for a clean git history and to keep `git bisect` happy (i.e., ideally there should not any build failures or failed test cases in any commit, even if the very next commit fixes the problem).

**Comment by bjorng** [comment]
I still see three failed cases in STDLIB:

```
erl_scan_SUITE:otp_7810/1
erl_scan_SUITE:otp_10302/1
tar_SUITE:unicode/1

```

**Comment by bjorng** [comment]
(I accidentally pushed an update of the primary bootstrap. You should remove that commit.)

**Comment by bjorng** [comment]
Can I suggest that you modify your test case to something like this?

```
list_to_utf8_atom(Config) when is_list(Config) ->
    'hello' = atom_roundtrip("hello"),
    'こんにちは' = atom_roundtrip("こんにちは"),

    %% Test all edge cases.
    _ = atom_roundtrip([16#80]),
    _ = atom_roundtrip([16#7F]),
    _ = atom_roundtrip([16#FF]),
    _ = atom_roundtrip([16#100]),
    _ = atom_roundtrip([16#7FF]),
    _ = atom_roundtrip([16#800]),
    _ = atom_roundtrip([16#D7FF]),
    atom_badarg([16#D800]),
    atom_badarg([16#DFFF]),
    _ = atom_roundtrip([16#E000]),
    _ = atom_roundtrip([16#FFFF]),
    _ = atom_roundtrip([16#1000]),
    _ = atom_roundtrip([16#10FFFF]),
    atom_badarg([16#110000]),
    ok.

atom_roundtrip(String) ->
    Atom = list_to_atom(String),
    Atom = list_to_existing_atom(String),
    String = atom_to_list(Atom),
    Atom.

atom_badarg(String) ->
    {'EXIT',{badarg,_}} = (catch list_to_atom(String)),
    {'EXIT',{badarg,_}} = (catch list_to_existing_atom(String)),
    ok.

```

**Comment by bjorng** [comment]
> Finally, during the weekend I took a look at most list_to_atom/atom_to_list calls in OTP and the ones that have called my attention are the ones in the compiler. I will take a double look at the compiler during this week and let you know if I find out anything suspicious.

Could write something about it in the commit message? That is, mention what you have looked for when looking at each call. That would help our own review and could also be useful five years later when someone investigates some bug that could possibly be connected with this change.

**Comment by josevalim** [comment]
I have squashed everything together and pushed your new tests. I will look at the failing stdlib tests and the list_to_atom changes later today.

**Comment by josevalim** [comment]
@bjorng I have fixed the erl_scan_SUITE. The tests were just outdated because they were expecting unicode atoms to fail. :)

Unfortunately, the tar_SUITE does not fail locally. Since it is using the filesystem, maybe it is something OS related? Which OS did the test fail on?

**Comment by josevalim** [comment]
I have done another look through all `list_to_atom/1` in the codebase. Most of the usages seem to be "safe", i.e. we are now allowing terms and constructs we didn't before.

For example, xmerl can now handle tag and attributes names in Unicode. Most of the compiler tool chain and related tools, such as the compiler itself, cover, syntax_tools is using list_to_atom to build tokens or to rename them. For example, to add `_` or `$` to a variable name.

The majority of those cases are fine except the ones listed below as questions:

  * `merl.erl` currently calls `list_to_atom(string:to_lower(atom_to_list(VarName)))`. This can be an issue because `string:to_lower/1` works only on Latin1. However, variable names can only be Latin1. So it is currently safe. Do we care? Should we add another TODO to the module (it has a list of them at the top)?

  * slave.erl uses `list_to_atom/1` on the node name. Should we check if we can have UTF-8 atoms as node names?

  * escript.erl converts the filename into a module name. Should we check if we can have UTF-8 atoms as module names and that escripts with such module names work accordingly?

  * epp.erl calls `code:lib_dir(list_to_atom(App))`. Should we check if we can have UTF-8 atoms as application names and that it will match the proper application directory name?

The last two are particularly tricky because it depends on the OS. I can write a test for Mac but unsure on how it will work on the other operating systems.

**Comment by bjorng** [comment]
> Unfortunately, the tar_SUITE does not fail locally. Since it is using the filesystem, maybe it is something OS related? Which OS did the test fail on?

Linux (Ubuntu 14.04). I will investigate this problem.

**Comment by bjorng** [comment]
> merl.erl currently calls list_to_atom(string:to_lower(atom_to_list(VarName))). This can be an issue because string:to_lower/1 works only on Latin1. However, variable names can only be Latin1. So it is currently safe. Should we add a note to the code?

Yes, that can be useful in case we start allowing Unicode variables.

> slave.erl uses list_to_atom/1 on the node name. Should we check if we can have UTF-8 atoms as node names? EDIT: non ascii characters on the node name fail miserably on this branch and master. I believe there is nothing to be done here.

How miserable? Would it better to crash in a controlled way instead of creating an Unicode atom and letting it crash latter?

> escript.erl converts the filename into a module name. Should we check if we can have UTF-8 atoms as module names and that escripts with such module names work accordingly? EDIT: escript with unicode names just work on Mac. I suggest to include tests and officially support this.

> epp.erl calls code:lib_dir(list_to_atom(App)). Should we check if we can have UTF-8 atoms as application names and that it will match the proper application directory name? EDIT: this seems to work just fine on Mac, I would recommend writing a test too.

I agree, but I have not discussed it with anyone else in the OTP team yet. We will also have to see whether it will cause any major problems for Linux and/or Windows.

**Comment by vladdu** [comment]
Regarding UTF-8 in file paths, it is tricky business. Refer to https://bugs.erlang.org/browse/ERL-176 for similar issues for Java compilation. For extended Unicode characters that are also in Latin1, it may happen that they get encoded the "wrong" way and different subsystems may handle these differently.

TL;DR: we should test as extensively as possible.

**Comment by josevalim** [comment]
> How miserable? Would it better to crash in a controlled way instead of creating an Unicode atom and letting it crash latter?

Erlang 18 silently discards it for UTF-8:

```
$ erl -name 'こんにちは'
Erlang/OTP 18 [erts-7.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V7.1  (abort with ^G)
1> node().
nonode@nohost
```

but fails horribly for latin1:

```
$ erl -sname 'josé'
{error_logger,{{2016,11,16},{17,51,16}},"Protocol: ~tp: the name jos�@macbook-2 seems to be in use by another Erlang node",["inet_tcp"]}
{error_logger,{{2016,11,16},{17,51,16}},crash_report,[[{initial_call,{net_kernel,init,['Argument__1']}},{pid,<0.22.0>},{registered_name,[]},{error_info,{exit,{error,badarg},[{gen_server,init_it,6,[{file,"gen_server.erl"},{line,344}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}},{ancestors,[net_sup,kernel_sup,<0.10.0>]},{messages,[]},{links,[#Port<0.344>,<0.19.0>]},{dictionary,[{longnames,false}]},{trap_exit,true},{status,running},{heap_size,233},{stack_size,27},{reductions,789}],[]]}
{error_logger,{{2016,11,16},{17,51,16}},supervisor_report,[{supervisor,{local,net_sup}},{errorContext,start_error},{reason,{'EXIT',nodistribution}},{offender,[{pid,undefined},{id,net_kernel},{mfargs,{net_kernel,start_link,[['josé',shortnames]]}},{restart_type,permanent},{shutdown,2000},{child_type,worker}]}]}
{error_logger,{{2016,11,16},{17,51,16}},supervisor_report,[{supervisor,{local,kernel_sup}},{errorContext,start_error},{reason,{shutdown,{failed_to_start_child,net_kernel,{'EXIT',nodistribution}}}},{offender,[{pid,undefined},{id,net_sup},{mfargs,{erl_distribution,start_link,[]}},{restart_type,permanent},{shutdown,infinity},{child_type,supervisor}]}]}
{error_logger,{{2016,11,16},{17,51,16}},crash_report,[[{initial_call,{application_master,init,['Argument__1','Argument__2','Argument__3','Argument__4']}},{pid,<0.9.0>},{registered_name,[]},{error_info,{exit,{{shutdown,{failed_to_start_child,net_sup,{shutdown,{failed_to_start_child,net_kernel,{'EXIT',nodistribution}}}}},{kernel,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,134}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}},{ancestors,[<0.8.0>]},{messages,[{'EXIT',<0.10.0>,normal}]},{links,[<0.8.0>,<0.7.0>]},{dictionary,[]},{trap_exit,true},{status,running},{heap_size,376},{stack_size,27},{reductions,117}],[]]}
{error_logger,{{2016,11,16},{17,51,16}},std_info,[{application,kernel},{exited,{{shutdown,{failed_to_start_child,net_sup,{shutdown,{failed_to_start_child,net_kernel,{'EXIT',nodistribution}}}}},{kernel,start,[normal,[]]}}},{type,permanent}]}
{"Kernel pid terminated",application_controller,"{application_start_failure,kernel,{{shutdown,{failed_to_start_child,net_sup,{shutdown,{failed_to_start_child,net_kernel,{'EXIT',nodistribution}}}}},{kernel,start,[normal,[]]}}}"}
```

Master raises the error above readably on latin1 but I believe the reason is wrong:

```
$ erl -sname 'josé'
Protocol 'inet_tcp': the name jos�@macbook-2 seems to be in use by another Erlang node
```

and a correct message on utf-8:

```
$ erl -sname 'こんにちは'
2016-11-16 17:50:04 ~s~n
	"Bad value on output port 'tcp_inet'\n"
Protocol 'inet_tcp': register/listen error: einval
```

So master works nicely for UTF-8 but shows an incorrect error message for Latin-1. The incorrect error message is not related to this PR.

**Comment by josevalim** [comment]
> I agree, but I have not discussed it with anyone else in the OTP team yet. We will also have to see whether it will cause any major problems for Linux and/or Windows.

I am also OK if the answer is: "we won't support unicode escript and application names because of filesystems". However, I believe we should add at least a test that a module compiled directly from the erlang abstract format with unicode characters in the module name works. That's a pre-requisite for eventually supporting UTF-8 named escripts anyway.

**Comment by bjorng** [comment]
> Unfortunately, the tar_SUITE does not fail locally. Since it is using the filesystem, maybe it is something OS related? Which OS did the test fail on?

It was two separate bugs in the test case. TL;DR is that the first bug hid the second bug, except when run in an interactive Erlang shell (our daily builds runs with standard output redirected to a file). I have corrected the bugs in a separate branch and will test it in our daily builds before merging to `maint` and `master`.

**Comment by bjorng** [comment]
I have now included the branch in our daily builds.

If that goes well, we will start looking closer at your code.

**Comment by bjorng** [comment]
I have pushed two commits.

First there is a minimal correction for `code_SUITE:load_abs/1` so that it does not fail.

Second there is the elimination of this warning:

```
beam/erl_unicode.c: In function ‘binary_to_atom’:
beam/erl_unicode.c:1898:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
     Eterm a;
     ^
```

(We still build OTP on a few systems with C90 compilers.)

Please squash these commits and rebase on the latest master.

**Comment by bjorng** [comment]
By the way, there were no other failed test cases. :+1:

**Comment by josevalim** [comment]
Squashed and rebased.

**Comment by bjorng** [comment]
The failure in Travis is not entirely obvious. ;)

It happens when Travis attempts to run dialyzer. The first thing dialyzer will do is to compile some modules to native code. To compile a module to native code, `beam_disasm` will be run on the BEAM file. If the BEAM file does not an 'AtU8' chunk, disassembly will fail:

```
1> beam_disasm:file("sets.beam").
{error,beam_lib,{missing_chunk,'sets.beam',"AtU8"}}
```

In our daily builds, we did not notice the problem, because we had updated the primary bootstrap, and therefore all BEAM files had an 'AtU8' chunk.

**Comment by bjorng** [comment]
I have pushed a commit to fix the beam_disasm problem. In beam_lib, I have removed the `utf8_atoms` symbolic chunk type. Instead, `atoms` will always return the symbolic atom table regardless of the "raw" name for the atom table.

There is another commit that eliminates a dialyzer warning and a commit with more test cases for beam_lib.

If you agree to the changes, please squash the commits and rebase to the latest master.

**Comment by josevalim** [comment]
Squashed and rebased.

**Comment by bjorng** [comment]
I found and fixed another minor bug. Please squash and rebase on latest master.

**Comment by josevalim** [comment]
Done and done!

Btw the reducing memory during compilation I had to rebase on may help Elixir too. :+1:

**Comment by bjorng** [comment]
Can you please rebase again?

We will test the branch in our daily builds again to see that there are no remaining failed test cases.

**Comment by josevalim** [comment]
@bjorng done!

**Comment by bjorng** [comment]
This branch no longer seems to cause any problems in our daily builds. We are thinking about merging it soon. More work is probably needed in some applications, for example to ensure that "~ts" is used as format specifier when an atom is displayed, but we can do that in separate branches.

However, I think that the documentation should be updated in this branch to reflect the changes, at least the BIFs for the atom BIFs. Do you think you could update the documentation?

**Comment by josevalim** [comment]
I will push the docs soon. :) Two questions:

1. Regarding `~ts`, I understand why it is needed for binaries and char lists, but it shouldn't be necessary for atoms assuming that atoms do have the encoding they are written in? So I suggest making `~s` work regardless of the atom encoding. Or do you see reasons where this would be a bad idea™?

2. Should we explicitly document module names can now be UTF-8? If so, I would like to add a test to compile_SUITE where we compile Erlang Forms module with a UTF-8 atom as name.

**Comment by josevalim** [comment]
I have pushed a separate commit with docs. Once review is done, I can squash it all together.

**Comment by bjorng** [comment]
1. Making `~s` work for any atom is an interesting idea. I will discuss it with the team.

2. We have not reached a final decision yet. Our current thinking is that we will allow Unicode characters in module names, but we will not recommend it because of potential portability problems having to do with the file names (Linux does not enforce any particular encoding for file names, and their could be other potential pitfalls when moving files from one operating system to another). The reason that we will probably allow it is because those potential problem already exists for module names with Latin1 characters.

**Comment by josevalim** [comment]
> Our current thinking is that we will allow Unicode characters in module names, but we will not recommend it because of potential portability problems having to do with the file names

I agree. Let me know what are your decisions regarding 1 and 2 and I will gladly add tests and improve the docs.

**Comment by bjorng** [comment]
There may some time before we can reach a decision. We will merge this branch and do further corrections/improvement in separate pull requests.

Please squash the commits. When you have done it, we will run the branch once more in our daily builds and then merge it.

**Comment by josevalim** [comment]
Squashed and pushed.

**Comment by nox** [comment]
> do further corrections/improvement in separate pull requests.

Like renaming the chunk to 'JOSÉ', right?

**Comment by josevalim** [comment]
🎉

**Comment by KronicDeth** [comment]
Support for decompiling `AtU8` chunks has been added to IntelliJ Elixir in KronicDeth/intellij-elixir#777, which will be released in version `6.0.0`.
